### PR TITLE
Mac: Turn on 'gvfs prefetch --hydrate' 

### LINKED ISSUE #243: Mac: ReadFilesJob references kernel32
ReadFilesJob references kernel32.dll in attempt to do native file open/read, which results in an exception being thrown during `gvfs prefetch --files *.txt --hydrate <enlistment>`

### PR DESCRIPTION
**Pull request by nickgra**
Fixes #243 

Implement a native macOS solution for 'gvfs prefetch --hydrate' and enable associated functional tests.

### CODE REVIEW

File: GVFS/GVFS.Common/GVFSPlatform.cs [hunk 1]
```diff
@@ -64,6 +64,8 @@ public static void Register(GVFSPlatform platform)
             ITracer tracer,
             string lockPath);
 
+        public abstract bool OpenAndReadFile(string fileName);
```
Comment (wilbaker) [inline GVFS/GVFS.Common/GVFSPlatform.cs]: Naming suggestion: `TryReadFirstByte` or `TryReadFirstByteOfFile`.

File: GVFS/GVFS.Platform.Windows/WindowsPlatform.cs [hunk 2]
```diff
@@ -285,5 +287,61 @@ private static object GetValueFromRegistry(RegistryHive registryHive, string key
...
+                uint shareMode,
+                uint securityAttributes,
+                uint creationDisposition,
+                uint flagsAndAttributes,
+                int hemplateFile);
+
+            [DllImport("kernel32", SetLastError = true)]
+            private static extern unsafe bool ReadFile(
+                SafeFileHandle file,
+                void* buffer,
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Windows/WindowsPlatform.cs]: Can we use `[Out] byte []` here as well and avoid having to add `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>`?

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs [hunk 3]
```diff
@@ -140,5 +141,41 @@ public override bool IsGitStatusCacheSupported()
...
+
+                if (success == -1)
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            private static unsafe int ReadOneByte(int fileDescriptor)
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Mac/MacPlatform.cs]: Maybe make this `TryReadOneByte` and have it return a bool?

It looked a bit odd to see:

```int success = ReadOneByte(fileDescriptor);```

When `ReadOneByte` is returning the number of bytes read.

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs [hunk 4]
```diff
@@ -140,5 +141,41 @@ public override bool IsGitStatusCacheSupported()
...
+            {
+                byte[] buffer = new byte[1];
+                return Read(fileDescriptor, buffer, 1);
+            }
+
+            [DllImport("libc", EntryPoint = "open", SetLastError = true)]
+            private static extern int Open(string path, int flag);
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Mac/MacPlatform.cs]: Not something for this PR, but at some point we might want to move all the native interop calls to a common class in `MacPlatform`

File: GVFS/GVFS.Platform.Windows/WindowsPlatform.cs [hunk 5]
```diff
@@ -285,5 +292,57 @@ private static object GetValueFromRegistry(RegistryHive registryHive, string key
...
+
+                return false;
+            }
+
+            private static SafeFileHandle Open(string fileName)
+            {
+                return CreateFile(fileName, GenericRead, (uint)(FileShare.ReadWrite | FileShare.Delete), 0, OpenExisting, 0, 0);
+            }
+
+            private static unsafe bool ReadOneByte(SafeFileHandle handle)
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Windows/WindowsPlatform.cs]: Name here too: `TryReadOneByte`

File: GVFS/GVFS.UnitTests/Mock/Common/MockPlatform.cs [hunk 6]
```diff
@@ -114,6 +114,11 @@ public override bool IsGitStatusCacheSupported()
         public override FileBasedLock CreateFileBasedLock(PhysicalFileSystem fileSystem, ITracer tracer, string lockPath)
         {
             return new MockFileBasedLock(fileSystem, tracer, lockPath);
-        }
+        }
+
+        public override bool OpenAndReadFile(string fileName)
+        {
+            throw new NotImplementedException();
```
Comment (sanoursa) [inline GVFS/GVFS.UnitTests/Mock/Common/MockPlatform.cs]: NotSupportedException

File: GVFS/GVFS.Common/GVFSPlatform.cs (hunk 1 above)
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSPlatform.cs]: I would move this to `IPlatformFileSystem`, and rename to `ReadOneByte`

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs (hunk 3 above)
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacPlatform.cs]: Does this actually need to be unsafe?

File: GVFS/GVFS.Common/GVFSPlatform.cs (hunk 1 above)
Comment (sanoursa) [reply GVFS/GVFS.Common/GVFSPlatform.cs]: Or `TryReadOneByte` as William suggested.

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs (hunk 4 above)
Comment (nickgra) [reply GVFS/GVFS.Platform.Mac/MacPlatform.cs]: I agree, I was thinking the same thing while working on this. Want me to open a work item for it?

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs [hunk 7]
```diff
@@ -6,7 +6,8 @@
 using System.Collections.Generic;
 using System.IO;
 using System.IO.Pipes;
-
+using System.Runtime.InteropServices;
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacPlatform.cs]: Drop this?

File: GVFS/GVFS.Platform.Windows/WindowsPlatform.cs [hunk 8]
```diff
@@ -4,12 +4,14 @@
 using GVFS.Common.Tracing;
 using GVFS.Platform.Windows.DiskLayoutUpgrades;
 using Microsoft.Win32;
+using Microsoft.Win32.SafeHandles;
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Windows/WindowsPlatform.cs]: Drop these changes too

File: GVFS/GVFS.Platform.Mac/MacFileSystem.cs [hunk 9]
```diff
@@ -36,6 +36,42 @@ public void ChangeMode(string path, int mode)
...
+        }
+
+        public bool TryReadFirstByteOfFile(string fileName)
+        {
+            return NativeFileReader.TryReadFirstByteOfFile(fileName);
+        }
+
+        private class NativeFileReader
+        {
+            // #define O_RDONLY     0x0000
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacFileSystem.cs]: I would drop this comment

File: GVFS/GVFS.Common/GVFSPlatform.cs [hunk 10]
```diff
@@ -63,7 +63,7 @@ public static void Register(GVFSPlatform platform)
             PhysicalFileSystem fileSystem,
             ITracer tracer,
             string lockPath);
-
+
```
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSPlatform.cs]: Drop this

File: GVFS/GVFS.Platform.Mac/MacPlatform.cs [hunk 11]
```diff
@@ -6,7 +6,7 @@
 using System.Collections.Generic;
 using System.IO;
 using System.IO.Pipes;
-
+
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacPlatform.cs]: Drop this change

File: GVFS/GVFS.Platform.Windows/WindowsPlatform.cs [hunk 12]
```diff
@@ -275,7 +275,7 @@ public override bool IsGitStatusCacheSupported()
         public override bool TryGetGVFSEnlistmentRoot(string directory, out string enlistmentRoot, out string errorMessage)
         {
             return WindowsPlatform.TryGetGVFSEnlistmentRootImplementation(directory, out enlistmentRoot, out errorMessage);
-        }
+        }
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Windows/WindowsPlatform.cs]: This and the next file are both space-only changes, drop them

File: GVFS/GVFS.Platform.Mac/MacFileSystem.cs [hunk 13]
```diff
@@ -36,6 +36,41 @@ public void ChangeMode(string path, int mode)
...
+
+            public static bool TryReadFirstByteOfFile(string fileName)
+            {
+                int fileDescriptor = Open(fileName, ReadOnly);
+                return TryReadOneByte(fileDescriptor);
+            }
+
+            private static bool TryReadOneByte(int fileDescriptor)
+            {
+                byte[] buffer = new byte[1];
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacFileSystem.cs]: We'll actually need to fix this pointer before passing it over to the native code. Otherwise the risk is that the GC can move this memory while the native code is running, and it will write to a now invalid pointer

File: GVFS/GVFS.Platform.Windows/WindowsPlatform.cs (hunk 2 above)
Comment (sanoursa) [reply GVFS/GVFS.Platform.Windows/WindowsPlatform.cs]: FYI we need to use fixed not just for the type conversion, but also to ensure the GC doesn't move the memory. So we're stuck with the unsafe keywords here.

File: GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs [hunk 14]
```diff
@@ -32,6 +35,63 @@ public void ChangeMode(string path, int mode)
...
+                uint creationDisposition,
+                uint flagsAndAttributes,
+                int hemplateFile);
+
+            [DllImport("kernel32", SetLastError = true)]
+            private static extern unsafe bool ReadFile(
+                SafeFileHandle file,
+                [Out] byte* buffer,
+                int numberOfBytesToRead,
+                [Out] int numberOfBytesRead,
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: How did this translate to an int? Shouldn't it be a pointer?

Comment (sanoursa) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: I think what's happening is that the type is not translated correctly here, but because the value you're passing in happens to be initialized to 0, the native code is probably seeing this as a null pointer and ignoring it. But anything other than a 0 will probably cause a crash.

Comment (sanoursa) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: And I can see now that this is the pre-existing Windows code, and I may have been the person who wrote this incorrect code :-)

File: GVFS/GVFS.Platform.Mac/MacFileSystem.cs [hunk 15]
```diff
@@ -36,12 +36,50 @@ public void ChangeMode(string path, int mode)
...
+                int fileDescriptor = Open(fileName, ReadOnly);
+                return TryReadOneByte(fileDescriptor);
+            }
+
+            private static unsafe bool TryReadOneByte(int fileDescriptor)
+            {
+                byte[] buffer = new byte[1];
+                fixed (byte* p = buffer)
+                {
+                    int resultCode = Read(fileDescriptor, p, 1);
```
Comment (sanoursa) [inline GVFS/GVFS.Platform.Mac/MacFileSystem.cs]: The return is actually `numBytes`, not `resultCode`, right? It just so happens that -1 is a special value.

File: GVFS/GVFS.Platform.Windows/GVFS.Platform.Windows.csproj [hunk 16]
```diff
@@ -12,6 +12,7 @@
     <AssemblyName>GVFS.Platform.Windows</AssemblyName>
     <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Windows/GVFS.Platform.Windows.csproj]: Is this change still needed?

File: GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs [hunk 17]
```diff
@@ -32,6 +35,59 @@ public void ChangeMode(string path, int mode)
...
+
+            private static SafeFileHandle Open(string fileName)
+            {
+                return CreateFile(fileName, GenericRead, (uint)(FileShare.ReadWrite | FileShare.Delete), 0, OpenExisting, 0, 0);
+            }
+
+            private static unsafe bool ReadOneByte(SafeFileHandle handle, byte[] buffer)
+            {
+                int bytesRead = 0;
+                return ReadFile(handle, buffer, 1, &bytesRead, 0);
```
Comment (wilbaker) [inline GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: Why are you passing in `&bytesRead`?  If you just pass in `bytesRead` there should be no need to mark this `unsafe`.

Comment (wilbaker) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: Oh sorry, I missed that this was for `int* numberOfBytesRead`

Comment (wilbaker) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: If you change the `DllImport` from:

`[Out] int* numberOfBytesRead,`

to 

`ref int numberOfBytesRead,`

You can drop the `unsafe`

Comment (wilbaker) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: See example here (the second parameter)

https://docs.microsoft.com/en-us/dotnet/framework/interop/marshaling-different-types-of-arrays

*C API*
```
int TestRefArrayOfInts(int** ppArray, int* pSize);  
```

*DllImport in C#*
```
   [ DllImport( "..\\LIB\\PinvokeLib.dll" )]
   public static extern int TestRefArrayOfInts( ref IntPtr array, 
      ref int size );
```

Comment (nickgra) [reply GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs]: Thanks!
